### PR TITLE
Ebpf xdp 5.0 v2.12

### PR DESCRIPTION
**Pull request by regit**
Rework of #3941 fixing the pinned maps.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:

Describe changes:
- Fix pinned maps
- Add counter for captured packets seen by local bypass
- Address comments of #3941 

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output:
- PR regit: https://buildbot.openinfosecfoundation.org/builders/regit/builds/471
- PR regit-pcap: https://buildbot.openinfosecfoundation.org/builders/regit-pcap/builds/252

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -1773,8 +1773,8 @@
         if test "$LIBBPF" = "no"; then
             echo
             echo "   libbpf library and development headers not found but"
-            echo "   but needed to use eBPF code. It can be found in the"
-            echo "   Linux kernel tree under tools/lib/bpf"
+            echo "   but needed to use eBPF code. It can be found in at"
```
Comment (norg) [inline configure.ac]: double **but** end of first and beginning of last line.

File: configure.ac [hunk 2]
```diff
@@ -1773,8 +1773,8 @@
...
+            echo "   https://github.com/libbpf/libbpf"
```
Comment (norg) [inline configure.ac]: what builtin version by kernel/distributions is also enough? maybe at this here or at the docs

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 3]
```diff
@@ -186,6 +195,9 @@ filter must expose `flow_table_v4` and `flow_table_v6` per CPU array maps with s
 as the one available in `bypass_filter.c`. These two maps will be accessed and
 maintained by Suricata to handle the lists of flow to bypass.
 
+If you are not using vlan tracking (``vlan.use-for-tracking`` set to false in suricata.yaml) then you have also to set
```
Comment (norg) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: then you also have to

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 4]
```diff
@@ -249,13 +261,26 @@ also use the ``/etc/suricata/ebpf/xdp_filter.bpf`` (in our example TCP offloadin
     bypass: yes
     use-mmap: yes
     ring-size: 200000
+    # Uncomment the following if you are using hardware XDP with
+    # a card like Netronome
+    # use-percpu-hash: no
```
Comment (norg) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: so the default if not uncommented is yes or ist it just unset?

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 5]
```diff
@@ -249,13 +261,26 @@ also use the ``/etc/suricata/ebpf/xdp_filter.bpf`` (in our example TCP offloadin
...
 
 
 XDP bypass is compatible with AF_PACKET IPS mode. Packets from bypassed flows will be send directly 
-from one card to the second card without going by the kernel network stack.
+from one card to the second card without going by the ker nel network stack.
```
Comment (norg) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: kernel without a space :)

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 6]
```diff
@@ -249,13 +261,26 @@ also use the ``/etc/suricata/ebpf/xdp_filter.bpf`` (in our example TCP offloadin
...
 
-Setup symmetric hashing on the NIC
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+If you are using hardware XDP offload you may have to set ``use-percpu-hash`` to false and
+build and install the XDP filter file after setting ``USE_PERCPU_HASH`` to 0.
+
+In the XDP filter file, you can set ``ENCRYPTED_TLS_BYPASS`` to 1 if you want to bypass
+the encrypted TLS 1.2 packetsin the eBPF code. Be aware that this will mean that Suricata will
```
Comment (norg) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: packets [[:space:]] in

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 7]
```diff
@@ -249,13 +261,26 @@ also use the ``/etc/suricata/ebpf/xdp_filter.bpf`` (in our example TCP offloadin
...
+be blind on packets on port 443 with the correct pattern.
+
+If you are not using vlan tracking (``vlan.use-for-tracking`` set to false in suricata.yaml) then you have also to set
```
Comment (norg) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: "then you also have to set"

and it's a duplicate of the prior chapter?

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 8]
```diff
@@ -350,11 +375,160 @@ Confirm you have the XDP filter engaged in the output (example)::
  ...
  ...
 
+Pinned maps usage
+-----------------
+
+Pinned maps stay attached to the system if the creating process disappears and
+they can also be accessed by external tools. In Suricata bypass case, this can be
+used to keep active bypassed flow tables so Suricata is not hit by previsouly bypassed flows when
```
Comment (norg) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: tables, so Suricata

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 9]
```diff
@@ -350,11 +375,160 @@ Confirm you have the XDP filter engaged in the output (example)::
...
+restarting. In the socket filter case, this can be used to maintain a map from tools outside
+of Suricata.
+
+To used pinned maps, you first have to mount the `bpf` pseudo filesystem ::
```
Comment (norg) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: To use

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 10]
```diff
@@ -350,11 +375,160 @@ Confirm you have the XDP filter engaged in the output (example)::
...
+
+  sudo mount -t bpf none /sys/fs/bpf
+
+You can also add to you `/etc/fstab` ::
```
Comment (norg) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: your

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 11]
```diff
@@ -350,11 +375,160 @@ Confirm you have the XDP filter engaged in the output (example)::
...
+will pin them under the name `suricata-$IFACE_NAME-$MAP_NAME`.
+
+To activate pinned maps for a interface, set `pinned-maps` to `true` in the `af-packet`
+configuration of this interface ::
+
+  - interface: eth3
+    pinned-maps: true
+
+This option can be used to expose the maps of a socket filter to other processes.
+This allows for example, the external handling of a blacklist or white list of
```
Comment (norg) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: whitelist

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 12]
```diff
@@ -350,11 +375,160 @@ Confirm you have the XDP filter engaged in the output (example)::
...
+for, that will be used to check for the presence of the XDP filter ::
+
+  - interface: eth3
+    pinned-maps: true
+    pinned-maps-name: ipv4_drop
+    xdp-filter-file: /etc/suricata/ebpf/xdp_filter.bpf
+
+If XDP bypass is used in IPS mode stopping Suricata will trigger an interruption in the traffic.
+To fix that, the provided XDP filter `xdp_filter.bpf` is containing a map that will trigger
+a global bypass if set to 1. You need to use `pinned-maps` to benefit of this feature.
```
Comment (norg) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: from this feature

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 13]
```diff
@@ -350,11 +375,160 @@ Confirm you have the XDP filter engaged in the output (example)::
...
+
+Hardware bypass with Netronome
+------------------------------
+
+Netronome cards support hardware bypass. In this case the eBPF code is running in the card
+itself. This introduces some architectural differences compared to driver mode and the configuration
+and eBPF filter need to be updated.
+
+On eBPF side, as of Linux 4.19 CPU maps and interfaces redirect are not supported and these features
+need to be disabled. By architecture, per CPU hash should not be used and have to be disabled.
```
Comment (norg) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: has to be disabled

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 14]
```diff
@@ -350,11 +375,160 @@ Confirm you have the XDP filter engaged in the output (example)::
...
+    xdp-mode: hw
+    no-percpu-hash: true
+
+The load  balancing will be done on IP pairs inside the eBPF code, so
+using `cluster_qm` as cluster type is a good idea ::
+
+    cluster-type: cluster_qm
+
+As of Linux 4.19, the number of threads must be a power of 2. So set
+`threads` variable interface of the `af-packet` interface to a power
```
Comment (norg) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: the first interface isn't necessary

File: doc/userguide/capture-hardware/ebpf-xdp.rst [hunk 15]
```diff
@@ -350,11 +375,160 @@ Confirm you have the XDP filter engaged in the output (example)::
...
+     "pkts": 56529854964,
+     "drop": 932328611,
+     "bypassed": 1569467248,
+     "invalid-checksums": 0
+   },
+   "return": "OK"
+ }
+
+``iface-bypassed-stats`` command will return the number of elements in IPv4 and IPv6 flow tables for
+each interfaces ::
```
Comment (norg) [inline doc/userguide/capture-hardware/ebpf-xdp.rst]: each interface

File: configure.ac (hunk 2 above)
Comment (regit) [reply configure.ac]: We got this in the doc:
```
Suricata XDP code has been tested with 4.13.10 but 4.15 or later is necessary to use all
features like the CPU redirect map.
```

File: doc/userguide/capture-hardware/ebpf-xdp.rst (hunk 7 above)
Comment (regit) [reply doc/userguide/capture-hardware/ebpf-xdp.rst]: Yes but you need it in both cases.

File: doc/userguide/capture-hardware/ebpf-xdp.rst (hunk 11 above)
Comment (regit) [reply doc/userguide/capture-hardware/ebpf-xdp.rst]: Switching to politically correct with accept list and block list.

**Comment by regit** [comment]
I'm gonna resubmit with amended doc.

**Pull request closed without merging**